src/mpid/ch4/shm/xpmem/xpmem_recv.h

```diff
@@ -9,71 +9,347 @@
+
+    DL_FOREACH_SAFE(*queue, cur_obj, next) {
+        if (cur_obj->counter_ptr->free_flag) {
+            DL_DELETE(*queue, cur_obj);
+            MPIR_Handle_obj_free(&MPIDI_XPMEM_cnt_mem, (void *) cur_obj->counter_ptr);
+            MPL_free(cur_obj);
+        }
+    }
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_XPMEM_COLLECT_CNT_OBJ);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_handle_lmt_coop_recv(uint64_t src_offset,
+                                                              size_t src_data_sz, uint64_t sreq_ptr,
+                                                              int src_lrank, MPIR_Comm * comm,
+                                                              MPIR_Request * rreq)
 {
     int mpi_errno = MPI_SUCCESS;
+    int dt_true_lb;
     MPIDI_XPMEM_seg_t *seg_ptr = NULL;
-    void *attached_sbuf = NULL;
+    void *src_buf = NULL;
+    void *dest_buf = NULL;
     size_t data_sz, recv_data_sz;
-    MPIDI_SHM_ctrl_hdr_t ack_ctrl_hdr;
-    MPIDI_SHM_ctrl_xpmem_send_lmt_ack_t *slmt_ack_hdr = &ack_ctrl_hdr.xpmem_slmt_ack;
+    int fin_type;
+    MPIDI_SHM_ctrl_hdr_t ctrl_hdr;
+    MPIDI_SHM_ctrl_xpmem_send_lmt_cts_t *slmt_cts_hdr = &ctrl_hdr.xpmem_slmt_cts;
 
-    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_HANDLE_LMT_RECV);
-    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_HANDLE_LMT_RECV);
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_XPMEM_HANDLE_LMT_COOP_RECV);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_XPMEM_HANDLE_LMT_COOP_RECV);
 
-    mpi_errno = MPIDI_XPMEM_seg_regist(src_lrank, src_data_sz, (void *) src_offset,
-                                       &seg_ptr, &attached_sbuf);
-    MPIR_ERR_CHECK(mpi_errno);
+    MPIDI_Datatype_check_size_lb(MPIDIG_REQUEST(rreq, datatype), MPIDIG_REQUEST(rreq, count),
+                                 data_sz, dt_true_lb);
 
-    MPIDI_Datatype_check_size(MPIDIG_REQUEST(rreq, datatype), MPIDIG_REQUEST(rreq, count), data_sz);
+    /* Data truncation checking */
+    recv_data_sz = MPL_MIN(src_data_sz, data_sz);
     if (src_data_sz > data_sz)
         rreq->status.MPI_ERROR = MPI_ERR_TRUNCATE;
 
-    /* Copy data to receive buffer */
-    recv_data_sz = MPL_MIN(src_data_sz, data_sz);
-    mpi_errno = MPIR_Localcopy(attached_sbuf, recv_data_sz,
-                               MPI_BYTE, (char *) MPIDIG_REQUEST(rreq, buffer),
-                               MPIDIG_REQUEST(rreq, count), MPIDIG_REQUEST(rreq, datatype));
+    /* XPMEM internal info */
+    slmt_cts_hdr->dest_offset = (uint64_t) MPIDIG_REQUEST(rreq, buffer) + dt_true_lb;
+    slmt_cts_hdr->data_sz = recv_data_sz;
+    slmt_cts_hdr->dest_lrank = MPIDI_XPMEM_global.local_rank;
+
+    /* Receiver replies CTS packet */
+    slmt_cts_hdr->sreq_ptr = sreq_ptr;
+    slmt_cts_hdr->rreq_ptr = (uint64_t) rreq;
+
+    /* free counter object that were cached in MPIDI_XPMEM_global.cnt_free_queue */
+    MPIDI_XPMEM_collect_cnt_obj(&MPIDI_XPMEM_global.cnt_free_queue);
+
```

Comment:
`MPIDI_XPMEM_collect_cnt_obj` is specifically used to cleanup `cnt_free_queue`. Thus, you do need to specify the queue as an input.

Besides, the function name `MPIDI_XPMEM_cleanup_free_cnts` might be more straightforward?